[clang] improve class type sugar preservation in pointers to members

mizvekov opened the pull request:
This changes the MemberPointerType representation to use a NestedNameSpecifier instead of a Type to represent the class.

Since the qualifiers are always parsed as nested names, there was an impedance mismatch when converting these back and forth into types, and this led to issues in preserving sugar.

The nested names are indeed a better match for these, as the differences which a QualType can represent cannot be expressed syntactically, and it also represents the use case more exactly, being either dependent or referring to a CXXRecord, unqualified.

This patch also makes the MemberPointerType able to represent sugar for a {up/downcast}cast conversion of the base class, although for now the underlying type is canonical, as preserving the sugar up to that point requires further work.

As usual, includes a few drive-by fixes in order to make use of the improvements, and removing some duplications, for example CheckBaseClassAccess is deduplicated from across SemaAccess and SemaCast.

Endilll reviewed (commented):
Changes to C++ DR tests look good.
> FIXME: why diagnostic says just `Y` and not `cwg794::Y`, like `cwg794::X`?

This FIXME resolved in the opposite direction I anticipated back when I wrote it, but that's fine, as long as we print both types in a consistent manner.

erichkeane reviewed (commented):
This patch ended up being pretty massive... ANYTHING we can do to reduce the size of this would be very appreciated.

I DID NOT do more than a cursory look over the tests, so if someone else could do some work to make sure there isn't anything awkward/weird happening, I'd appreciate it.

Otherwise, I have some minor concerns/issues, but the code itself seems fine.

mizvekov commented on clang/include/clang/AST/Type.h:
```diff
@@ -3524,14 +3524,16 @@ class MemberPointerType : public Type, public llvm::FoldingSetNode {
-  /// RecordType, but could be a typedef or a template parameter too.
-  const Type *Class;
+  /// CXXRecordType, but could be a typedef or a template parameter too.
+  NestedNameSpecifier *Qualifier;
```
Note that we use the 'Qualifier' name consistently to refer to the NestedNameSpecifier, including the 'getQualifier' accessor, and this goes across not only Types (ElaboratedType, DependentNameType, DependentTemplateSpecialziationType), as well as expressions (DeclRefExpr), and even TemplateName, as well (QualifiedTemplateName).

A simple `git grep getQualifier clang/include/` will show that in fact most hits are for the NestedNameSpecifier, and the Type qualifiers are in the minority.

We even have generic functions which are templated on the node and assume the 'getQualifier' spelling works across the board.

In any case, a patch to change this consistently would be massive on it's own.

erichkeane replied:
ah, got it, it is an unfortunate overload of the word.  I'm shocked that I'd not noticed it in this use before.

mizvekov commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -10033,15 +10040,24 @@ class Sema final : public SemaBase {
                                  bool InOverloadResolution,
                                  QualType &ConvertedType);
 
+  enum class CheckMemberPointerConversionResult {
```
Yes Check is meaningful, because this goes into the function which just checks if the conversion is valid, but does not perform the conversion itself :)

erichkeane replied:
Right, it is the result of the check, but 'result' IMO implies that as well.  The word `Check` here to me doesn't really add anything to the names.

mizvekov replied:
Right but the name of the result type is unfortunately not scoped on the function, so I was afraid this type would show up on someone's autocomplete, and it would not be obvious to which function this is about.

In any case, I will give this some more thought.

mizvekov commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -13461,13 +13486,112 @@ static ElaboratedTypeKeyword getCommonTypeKeyword(const T *X, const T *Y) {
 
+static NestedNameSpecifier *getCommonNNS(ASTContext &Ctx,
+                                         NestedNameSpecifier *X,
+                                         NestedNameSpecifier *Y, bool Sugar) {
```
Yeah this is the convention we took on these functions, and LHS / RHS would not be accurate I think.
If we went with Something1 and Something2, instead of SomethingX and SomethingY, this would be another big patch to make consistent across the board.

I think X and Y for the parameters helps, instead of SomethingX, as it conveys its the kind the function operates on and returns.
For example, a getCommonType function gets the common types of X and Y, and getCommonNNS gets the common nested name of X and Y.

erichkeane replied:
I still think X and Y need better names, but could be talked out of it as convention (I see we're using it elsewhere here), but `Sugar` needs to find a better name.  Your description of its meaning doesn't show that implication at all.

mizvekov commented on clang/lib/AST/Type.cpp:
```diff
@@ -5172,8 +5179,28 @@ QualType::DestructionKind QualType::isDestructedTypeImpl(QualType type) {
+bool MemberPointerType::isSugared() const {
+  CXXRecordDecl *D1 = getMostRecentCXXRecordDecl(),
+                *D2 = getQualifier()->getAsRecordDecl();
+  return D1 != D2 && D1->getCanonicalDecl() != D2->getCanonicalDecl();
```
It's by the way the reason we don't use declaresSameEntity here, as one very annoying thing that function does is to return false even if both are null.

mizvekov commented:
> This patch ended up being pretty massive... ANYTHING we can do to reduce the size of this would be very appreciated.

It might be possible to split some stuff, but it would be hard to test it individually and avoid regressions in between the patches.

mizvekov commented:
A dependent PR had been merged meanwhile, and this was not rebased.

While this patch is still big, you might have been seeing way more stuff than was actually in the PR.

I just rebased, but have not addressed any comments yet.

mizvekov commented:
I had already done so, it's practically no change: https://llvm-compile-time-tracker.com/compare.php?from=cce9f360b50ca5a95d5cb550e27a8f6757423460&to=20841e444c0f8d48c6c9d34a77fe48164f7ae728&stat=instructions:u

mizvekov committed:
[clang] preserve class type sugar when taking pointer to member

This changes the MemberPointerType representation to use
a NestedNameSpecifier instead of a Type to represent the base
class.

Since the qualifiers are always parsed as nested names, there
was an impedance mismatch when converting these back and
forth into types, and this led to issues in preserving
sugar.

The nested names are indeed a better match for these,
as the differences which a QualType can represent cannot be
expressed syntatically, and they represent the use case more
exactly, being either dependent or referring to a CXXRecord,
unqualified.

This patch also makes the MemberPointerType able to represent
sugar for a {up/downcast}cast conversion of the base class,
although for now the underlying type is canonical, as preserving
the sugar up to that point requires further work.

As usual, includes a few drive-by fixes in order to make use
of the improvements.

mizvekov commented:
Done, in order to expedite review, I have split-off three patches from this:
* https://github.com/llvm/llvm-project/pull/131964
* https://github.com/llvm/llvm-project/pull/131965
* https://github.com/llvm/llvm-project/pull/131966

erichkeane commented:
> Done, in order to expedite review, I have split-off three patches from this:
> 
>     * [[clang] ASTContext: flesh out implementation of getCommonNNS #131964](https://github.com/llvm/llvm-project/pull/131964)
> 
>     * [[clang] NFC: Clear some uses of MemberPointerType::getClass #131965](https://github.com/llvm/llvm-project/pull/131965)
> 
>     * [[clang] NFC: Unify implementations of CheckMemberPointerConversion #131966](https://github.com/llvm/llvm-project/pull/131966)

Thanks!  Also, ooof, still 69 changed files.

erichkeane commented on clang/include/clang/AST/TypeLoc.h:
```diff
@@ -1355,7 +1355,7 @@ class BlockPointerTypeLoc : public PointerLikeTypeLoc<BlockPointerTypeLoc,
 
 struct MemberPointerLocInfo : public PointerLikeLocInfo {
-  TypeSourceInfo *ClassTInfo;
+  void *QualifierData = nullptr;
```
Could we instead use one of LLVM's pointer union types? 

Ah, urgh, I see we're inheriting this from `NestedNameSpecifierLoc`.  Its a shame we don't have a better 'thing' to store.  I'm OK leaving a 'fixme' here instead.

mizvekov replied:
This is actually not just an issue with NestedSpecifierLoc.

TypeLoc provides access to the opaque data through this void pointer, through the getOpaqueData method.
I added a comment there, but chances are low somebody is going to do something about it anytime soon :)

There are far worse issues just on TypeLoc itself which are still there from the initial implementation 15 years ago.

erichkeane reviewed (approved):
just want for 1 FIXME comment, else I've reviewed as well as I can and found nothing concerning.

erichkeane commented:
I'm still ok with this.

mizvekov commented:
Oof, looks like lldb needs updating too.
